Adding support for multiple contracts

matthewjacobson opened the pull request:
Adding support for multiple contracts. Had to change the .env file a bit - now passing all contract info as a single json string. Not a very elegant solution but I couldn't think of anything better. It would look like this for a plottables deployment:

```
{
    "0xa319c382a702682129fcbf55d514e61a16f97f9c": {
        "MINT_CONTRACT_ADDRESS": "0x463b8ced7d22a55aa4a5d69ef6a54a08aa0feb93",
        "MEDIA_URL": "https://plottables-mainnet.s3.amazonaws.com",
        "TOKEN_URL": "https://token.artblocks.io",
        "GENERATOR_URL": "https://generator.artblocks.io"
    },
    "0x18de6097ce5b5b2724c9cae6ac519917f3f178c0": {
        "MINT_CONTRACT_ADDRESS": "0xe6e728361b7c824cba64cc1e5323efb7a5bb65da",
        "MEDIA_URL": "https://plottables-flex-mainnet.s3.amazonaws.com",
        "TOKEN_URL": "https://token.artblocks.io",
        "GENERATOR_URL": "https://generator.artblocks.io"
    }
}
```

The TOKEN_URL and GENERATOR_URL per project is probably redundant but figured it wouldn't hurt to throw them in there.

Future Additions:
- Minting support for older contract versions - would probably require a new field in the REACT_APP_CONTRACT_INFO .env variable indicating the version when deploying for contracts spanning multiple versions.
- A User page listing all tokens owned by a wallet

Issues:
- On the /projects page the pagination is kind of broken - but only when trying to go to a page number lower than the one currently selected. Need to find a better way to get the total number of projects - currently using max project id which breaks when using multiple contracts.

matthewjacobson commented:
Added migration to latest versions of rainbowkit and wagmi to avoid the issue mentioned at https://github.com/rainbow-me/rainbowkit/issues/837

matthewjacobson commented:
The issue I mentioned earlier on the /projects page isn't actually related to pagination (as far as I can tell). Seems to be a bug in mui masonry as mentioned at https://stackoverflow.com/questions/75774800/how-to-stop-resizeobserver-loop-limit-exceeded-error-from-appearing-in-react-a

However I do still think using the max project id will be a problem

grantoesterling commented on sample.env:
```diff
@@ -1,8 +1,4 @@
-REACT_APP_MEDIA_URL=https://media.artblocks.io
-REACT_APP_TOKEN_URL=https://token.artblocks.io
-REACT_APP_GENERATOR_URL=https://generator.artblocks.io
+REACT_APP_CONTRACT_INFO={"0x99a9b7c1116f9ceeb1652de04d5969cce509b069":{"MINT_CONTRACT_ADDRESS":"0x706d6C6ef700a3c1C3a727f0c46492492E0A72b5","MEDIA_URL":"https://media.artblocks.io","TOKEN_URL":"https://token.artblocks.io","GENERATOR_URL":"https://generator.artblocks.io"}}
```
Could this be defined in a .json config file instead of the .env file ([a la artbot](https://github.com/ArtBlocks/artbot/blob/main/src/ProjectConfig/channels.json))? seems like that would be much more readable/sustainable, plus i don't think this info is confidential or anything

jakerockland replied:
++

Jbern16 replied:
yes please!

matthewjacobson replied:
Agreed this is a great idea. Addressed with [49e7b63](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/49e7b639877e81958340e0d58e1095c25ed2aaf2)

grantoesterling commented on src/components/Header.tsx:
```diff
@@ -19,11 +19,6 @@ const items = [
     enabled: true
-  },
-  {
-    label: "Mint",
```
Do you think this should be kept in as an example of a disabled / potential tab @jakerockland ?

jakerockland replied:
That seems reasonable to me. @Jbern16 what are your thoughts here?

Jbern16 replied:
Seems fine, generally a fan of more examples the better

matthewjacobson replied:
address with [59d2ffa](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/59d2ffa1d07b84eb86856625d2570d9849365f52)

grantoesterling commented on src/components/ProjectPreview.tsx:
```diff
@@ -24,7 +24,7 @@ const ProjectPreview = ({project, width=280, showDescription=false}: Props) => {
     <Box>
       <Box>
-        <Link href={`/project/${project.projectId}`} underline="hover">
+        <Link href={`/project/${project.id.split('-')[0]}/${project.projectId}`} underline="hover">
```
I think it makes sense to add the `contract` field to the `GetProject` query in `useProject.tsx` in order to do `project.contract.address` instead of the splitting here

matthewjacobson replied:
addressed with [cc69df2](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/cc69df2d9aca3b7a9ce1e7f4dfd30fbc722eca3f)

grantoesterling commented on src/components/TokenImage.tsx:
```diff
@@ -7,9 +7,10 @@ interface Props {
   return (
     <img
-      src={`${MEDIA_URL}/${tokenId}.png`}
+      src={`${mediaUrl}/${tokenId.split('-')[1]}.png`}
```
why does this split need to happen? is the tokenId that is passed in different now?

matthewjacobson replied:
addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

grantoesterling commented on src/components/TokenLive.tsx:
```diff
@@ -20,7 +20,8 @@ const TokenLive = ({tokenId, width, height}: Props) => {
   const [pollingAttempts, setPollingAttempts] = useState(0)
-  const endpoint = `${GENERATOR_URL}/${CORE_CONTRACT_ADDRESS?.toLowerCase()}/${tokenId}`
+  const generatorUrl = CONTRACT_INFO[tokenId.split('-')[0]].GENERATOR_URL
+  const endpoint = `${generatorUrl}/${tokenId.split('-')[0].toLowerCase()}/${tokenId.split('-')[1]}`
```
ditto the above, we can probs add the contract field to the GetToken query instead of splitting all these places, for readability reasons

matthewjacobson replied:
addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

grantoesterling commented on src/components/TokenView.tsx:
```diff
@@ -36,10 +35,10 @@ const TokenView = ({
         (
         <Box sx={{width: "100%", display: "flex", justifyContent: "flex-end"}}>
-          <Link href={`/token/${CORE_CONTRACT_ADDRESS}-${tokenId}`} sx={{fontSize: "14px", textDecoration: "none"}}>
+          <Link href={`/token/${tokenId.split('-')[0]}/${tokenId.split('-')[1]}`} sx={{fontSize: "14px", textDecoration: "none"}}>
```
ditto above on the splitting

matthewjacobson replied:
addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

grantoesterling commented on src/components/Tokens.tsx:
```diff
@@ -63,14 +63,14 @@ const Tokens = ({
           data.tokens.map(((token:Token) => (
             <Grid key={token.tokenId} item md={4} sm={6} xs={6}>
-              <Link href={`/token/${token.id}`}>
+              <Link href={`/token/${token.id.split('-')[0]}/${token.id.split('-')[1]}`}>
```
splitting

matthewjacobson replied:
addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

grantoesterling commented on src/hooks/useGeneratorPreview.tsx:
```diff
@@ -28,7 +28,8 @@ const useGeneratorPreview = (project: Project) => {
     try {
       const token = generateToken(Number(project.projectId))
-      const { data } = await axios.get(`${GENERATOR_URL}/${project.id}/${token.tokenId}/${token.hash}`)
+      const generatorUrl = CONTRACT_INFO[project.id.split('-')[0]].GENERATOR_URL
```
splitting

matthewjacobson replied:
addressed with [49e7b63](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/49e7b639877e81958340e0d58e1095c25ed2aaf2)

grantoesterling reviewed (dismissed):
looks good! just a few comments / questions :)

Jbern16 commented on src/components/ProjectDetails.tsx:
```diff
@@ -75,7 +76,7 @@ const ProjectDetails = ({ id }: Props) => {
             <Grid item md={8}>
               <TokenView
-                tokenId={token.tokenId}
+                tokenId={token.id}
```
don't think we need to change this and instead use the token.contract_address instead where we need the contract addy

matthewjacobson replied:
addressed with [7202d8b](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/7202d8bf8996b2b5c2c735f6bc4f98af38528f59)

matthewjacobson commented:
Thanks so much for the feedback! Going to start addressing some of your comments.

jakerockland commented:
Re-requesting review from your end @grantoesterling

grantoesterling commented on src/components/TokenDetails.tsx:
```diff
@@ -22,14 +22,18 @@ import useToken from "hooks/useToken"
+  const { loading, error, data } = useToken(`${contractAddress}-${id}`)
   const token = data?.token
+  const contractConfig = CONTRACT_INFO.filter(
+      x => x.CORE_CONTRACT_ADDRESS.toLowerCase() == contractAddress.toLowerCase()
```
nit: I think this should be `===`

matthewjacobson replied:
Addressed with [071be03](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/071be0363e6a4237b3733c943f00446b578e7f0c) in utils/contractInfoHelper.ts

grantoesterling commented on src/components/TokenDetails.tsx:
```diff
@@ -86,7 +91,8 @@ const TokenDetails = ({ id }: Props) => {
                 }}
                 onClick={() => {
-                  window.open(`${GENERATOR_URL}/${CORE_CONTRACT_ADDRESS?.toLowerCase()}/${token.tokenId}`)
+                  const generatorUrl = contractConfig[0].GENERATOR_URL
```
is this supposed to be hardcoded as 0?

grantoesterling replied:
I suppose it doesn't matter since the links are the same, but for robustness' sake should probably make this programatic depending on the token (in a world where someone might have 2 diff generator urls). A util function of some sort maybe where you pass in contractAddress and get generatorUrl/mediaUrl/etc?

Alternatively if generator link will ALWAYS be the same for everyone, maybe it is just its own config var instead of being in that contractConfig struct

matthewjacobson replied:
The 0 hardcoding was intentional (although probably bad practice) - I was using CONTRACT_INFO.filter() to get the proper contract config, and this returned an array for which there should have been only a single element. I have addressed this with [071be03](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/071be0363e6a4237b3733c943f00446b578e7f0c). Now using CONTRACT_INFO.find() to avoid the hardcoding.

grantoesterling commented on src/components/TokenDetails.tsx:
```diff
@@ -103,7 +109,8 @@ const TokenDetails = ({ id }: Props) => {
                 }}
                 onClick={() => {
-                  window.open(`${MEDIA_URL}/${token.tokenId}.png`)
+                  const mediaUrl = contractConfig[0].MEDIA_URL
```
ditto the above

grantoesterling replied:
Actually this one definitely shouldn't be hardcoded right? since media urls will be different?

Jbern16 replied:
media_url is this context I believe might be referencing our media proxy? However, you should be able to get the media_url from the tokens_metadata. `token.media_url`

matthewjacobson replied:
Address with [071be03](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/071be0363e6a4237b3733c943f00446b578e7f0c)

@grantoesterling yes the media url will be different for each contract - at least that has been my experience with plottables.

@Jbern16 I'm not seeing the media_url in the tokens_metadata but maybe I am missing something. Is this coming from the subgraph somewhere?

Jbern16 replied:
ah - my apologies, this is from our off chain api. We should be opening that up to the public soon, but perhaps until then we can hardcode the endpoints as it might streamline this logic - 

```
https://media-proxy.artblocks.io/0x54a6356244059d5a50b97200a928f19a3682b669/10
https://media-proxy-staging.artblocks.io/0xda62f67be7194775a75be91cbf9feedcc5776d4b/97000751
https://media-proxy-development.artblocks.io/0x5702797ff45fcb0a70eb6ae1e4563299dcfa9dd6/1

{contract_address}/{token_id}
```

grantoesterling commented on src/components/TokenImage.tsx:
```diff
@@ -1,15 +1,20 @@
   return (
     <img
-      src={`${MEDIA_URL}/${tokenId}.png`}
+      src={`${contractConfig[0].MEDIA_URL}/${tokenId}.png`}
```
ditto hardcoded index comment

matthewjacobson replied:
Address with [071be03](https://github.com/ArtBlocks/artblocks-engine-react/pull/10/commits/071be0363e6a4237b3733c943f00446b578e7f0c)

grantoesterling reviewed (approved):
LGTM other than the question about hardcoded indicies!

jakerockland commented:
Thank you very much @grantoesterling @Jbern16 for the thoughtful reviews here.

Thank you @matthewjacobson for sending along this PR !! 💜 🙏

matthewjacobson commented:
> Thank you very much @grantoesterling @Jbern16 for the thoughtful reviews here.
> 
> Thank you @matthewjacobson for sending along this PR !! 💜 🙏

Thank you! And thanks to @grantoesterling and @Jbern16 for the reviews!

Two other things we could maybe address before pulling:
1) Do any of you have thoughts on the best way to add support for the older V2 contracts? Would it make sense to add a new MintingInterfaceV2 component? And then do something like this in ProjectDetails.tsx:
```
{
  {
    "V3": <MintingInterfaceV3 ... />,
    "V2": <MintingInterfaceV2 ... />
  }
}[contractConfig?.VERSION]
```
2) Are you all experiencing the same issue I am on the /projects page? Resizing the page (or going from page 2 back to page 1 throws a "ResizeObserver loop limit exceeded" error. As I mentioned earlier this seems to be a [bug in mui masonry](https://stackoverflow.com/questions/75774800/how-to-stop-resizeobserver-loop-limit-exceeded-error-from-appearing-in-react-a). Should I update this to use mui grid?

Jbern16 commented:
@matthewjacobson 

> > Thank you very much @grantoesterling @Jbern16 for the thoughtful reviews here.
> > Thank you @matthewjacobson for sending along this PR !! 💜 🙏
> 
> Thank you! And thanks to @grantoesterling and @Jbern16 for the reviews!
> 
> Two other things we could maybe address before pulling:
> 
> 1. Do any of you have thoughts on the best way to add support for the older V2 contracts? Would it make sense to add a new MintingInterfaceV2 component? And then do something like this in ProjectDetails.tsx:
> 
> ```
> {
>   {
>     "V3": <MintingInterfaceV3 ... />,
>     "V2": <MintingInterfaceV2 ... />
>   }
> }[contractConfig?.VERSION]
> ```
> 

This looks like a solid pattern to me, we don't really have a prescribed way to do this, so this lgtm!

matthewjacobson commented:
Sorry for the Vercel stuff - I think that was my fault.

So I think I have the last two things figured out. It took a while for me to figure out the erc20 stuff for the V2 contracts but I think I got it sorted out. Please review and let me know if you have any questions!

jakerockland commented:
Staged deployment is looking good to me!! 🙌 🙌 

https://artblocks-engine-react-96lo14cc4-artblocks.vercel.app/projects

QQ: have you validated that the purchase flow is working as you'd expect with this changeset @matthewjacobson ?

matthewjacobson commented:
> Staged deployment is looking good to me!! 🙌 🙌
> 
> https://artblocks-engine-react-96lo14cc4-artblocks.vercel.app/projects
> 
> QQ: have you validated that the purchase flow is working as you'd expect with this changeset @matthewjacobson ?

Thanks for taking a look! I am currently in the process of validating the purchase flow - should have a few more changes pushed in a day or two. I am currently only able to test the MinterSetPriceV4 and MinterMerkleV5 minters as that is what has been deployed for the [plottables V3 contracts](https://github.com/ArtBlocks/artblocks-contracts/blob/32738da594e7b9d18e25011b1c7fefa4abb1bda9/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md). Happy to test others out if you want to deploy more on the plottables minter filter :)

jakerockland commented:
> > Staged deployment is looking good to me!! 🙌 🙌
> > https://artblocks-engine-react-96lo14cc4-artblocks.vercel.app/projects
> > QQ: have you validated that the purchase flow is working as you'd expect with this changeset @matthewjacobson ?
> 
> Thanks for taking a look! I am currently in the process of validating the purchase flow - should have a few more changes pushed in a day or two. I am currently only able to test the MinterSetPriceV4 and MinterMerkleV5 minters as that is what has been deployed for the [plottables V3 contracts](https://github.com/ArtBlocks/artblocks-contracts/blob/32738da594e7b9d18e25011b1c7fefa4abb1bda9/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md). Happy to test others out if you want to deploy more on the plottables minter filter :)

Fantastic 🙏 🙏 

I think that is perfect for now – no need to test with others unless you want to add support for others and test it for Plottables anyway, in which case would be happy to get you additional minters as-needed @matthewjacobson 

cc @ABDruid

matthewjacobson commented:
Hey @jakerockland - I have verified the purchase flows for the old V2 contract and the new MinterSetPriceV4 and MinterMerkleV5 minters. I included the original logic for MinterDAExpV4 as well but haven't tested it yet. Happy to run through that if I can get it deployed. I think the only other minters I would be interested in supporting for Plottables would be MinterHolderV4 and MinterSetPriceERC20V4 - but no pressure if there is no bandwidth for the deployment work!

Only other comment is related to MinterMerkleV5. I wrote it to hit an api endpoint (.env variable) in order to calculate the merkle proof. Currently this api is something that each partner would have to set up and configure. Two questions:
1) does this seem like a reasonable/generalized solution?
2) would this be something that AB could support natively in the future?

Otherwise let me know if you have any questions or if you want me to add a wallet to the allowlist for the MinterMerkleV5 test project or send some ERC20 tokens for the V2 ERC20 test project.

jakerockland commented:
> Hey @jakerockland - I have verified the purchase flows for the old V2 contract and the new MinterSetPriceV4 and MinterMerkleV5 minters. I included the original logic for MinterDAExpV4 as well but haven't tested it yet. Happy to run through that if I can get it deployed. I think the only other minters I would be interested in supporting for Plottables would be MinterHolderV4 and MinterSetPriceERC20V4 - but no pressure if there is no bandwidth for the deployment work!

Amazing ser – super appreciate you validating that logic, and no worries on the DA testing. In that case, would you mind updating the README to reflect that the supported minters to reflect the ones that you've tested? I think its OK to remove the EXPDA from that list for now until we get a chance to test on our end. 

Regarding getting you a MinterHolderV4 and MinterSetPriceERC20V4, I'm more than happy to do so! Especially given that you're helping improve the template here which is **super super** appreciated ser. 

Was just doing a bunch of deployments today anyway, so added yours to the list, you can find the full details here (note that you will need to allowlist them on your minterfilter yourself, as I couldn't automate this part given that we've already transferred ownership over to you): https://github.com/ArtBlocks/artblocks-contracts/blob/09efc513c829162c4e7a3444d90b3ef4995eb7b1/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md#minter-deployment

> Only other comment is related to MinterMerkleV5. I wrote it to hit an api endpoint (.env variable) in order to calculate the merkle proof. Currently this api is something that each partner would have to set up and configure. Two questions:
> 
> 1. does this seem like a reasonable/generalized solution?
> 2. would this be something that AB could support natively in the future?

I think this is a reasonable solution for now – would just ask that you add details to the README clarifying that.

That said, regarding the handling of this in the future (why I say "for now"), we do have plans in the near future to move to a world where there is a unified minter suite that can be used across all V3+ core contracts, meaning that there is a shared minterfilter and shared set of minters (concrete implementation wise, not just same contract code wise) which will make it such that the Art Blocks admin/artist experience can be used for things like minter configuration (including merkle proof minter management).

Once this is complete, I _think_ it should be relatively simple on our end to also provide the merkle proof calculation, but will need to talk more with @Jbern16 @yoshiwarab and @ryley-o on that one internally to get a sense of how we might expose that to partners.

tl;dr, we don't have specific immediate plans committed to how we'll approach that once our minter suite re-architecture work is completed, but its not off the table.

matthewjacobson commented:
Amazing! Thanks so much @jakerockland!

I'll get to work on adding support for those two new minter types and start working on updating the README

jakerockland commented:
> Amazing! Thanks so much @jakerockland!
> 
> I'll get to work on adding support for those two new minter types and start working on updating the README

Thank **YOU** so much ser! Sounds great to me 🙏 🙏 🙏

matthewjacobson commented:
Hey @jakerockland - apologies for being out of touch recently, I was having too much fun in Tokyo :) but I'm back now and ready to get back to work!

I saw your message over on #14 - I'll sync those changes in.

With respect to the two other minters you deployed:
1) Is there any documentation on how the MinterHolderV4 is supposed to work?
2) Any chance we can verify and publish the contract source code for MinterSetPriceERC20V4? Currently not able to interact with it on etherscan - https://goerli.etherscan.io/address/0xE79B6D89cc2126d8f1bEb128C748b25cc6D21420#code

jakerockland commented:
> Hey @jakerockland - apologies for being out of touch recently, I was having too much fun in Tokyo :) but I'm back now and ready to get back to work!

No worry at all – thank you so much @matthewjacobson 💜 🙏 

> 1. Is there any documentation on how the MinterHolderV4 is supposed to work?

Unfortunately we don't have _great_ documentation on this right now, but the tl;dr is that it is intended to be used to gate minting to the holders of another Engine or Flagship project (e.g. gate minting to only Squiggle holders).

More detailed breakdown can be found in the natspec: https://artblocks.github.io/artblocks-contracts/#/contracts/archive/minter-suite/Minters/MinterHolder/MinterHolderV3.sol:MinterHolderV3, though this is pretty verbose.

Happy to answer any questions though that you have to help clarify ser!  

> 2. Any chance we can verify and publish the contract source code for MinterSetPriceERC20V4? Currently not able to interact with it on etherscan - https://goerli.etherscan.io/address/0xE79B6D89cc2126d8f1bEb128C748b25cc6D21420#code

Whoops, my apology I did not realize that the verification initially failed.

Updated:

```
Successfully verified contract MinterSetPriceERC20V4 on Etherscan.
https://goerli.etherscan.io/address/0xE79B6D89cc2126d8f1bEb128C748b25cc6D21420#code
```

matthewjacobson commented:
Hi @jakerockland!

Okay I have added support for the MinterSetPriceERC20V4 and MinterHolderV4 minters. I can start working on the DA minters once you can get those deployed for me on goerli.

With respect to the upstream merge mentioned in #14 - to be honest I am a bit confused on the best way to go about it. I am definitely not a git wizard and I am getting a bit lost on what exactly I should do. Any guidance would be very much appreciated 💜

jakerockland commented:
> Okay I have added support for the MinterSetPriceERC20V4 and MinterHolderV4 minters. I can start working on the DA minters once you can get those deployed for me on goerli.

Amazing, will get two exponential DA minters (with and w/o settlement) deployed for you later today and post an update that has updated deployment logs!

> With respect to the upstream merge mentioned in #14 - to be honest I am a bit confused on the best way to go about it. I am definitely not a git wizard and I am getting a bit lost on what exactly I should do. Any guidance would be very much appreciated 💜

If you click on the "Use the command line" hyperlink that I believe you should see under the conflicts message, it should walk you through how to pull down the upstream changes into your branch, and then from there you should just resolve any conflicts by picking the changes that reflect your changeset and work as expected still with your updates.

<img width="424" alt="Screenshot 2023-05-22 at 10 30 03 AM" src="https://github.com/ArtBlocks/artblocks-engine-react/assets/8602661/d52f8f1e-0e91-4dc8-8b5a-c5eed3967c6c">

Let me know how I can help clarify further ser and am happy to help. 🙏

matthewjacobson commented:
> If you click on the "Use the command line" hyperlink that I believe you should see under the conflicts message, it should walk you through how to pull down the upstream changes into your branch, and then from there you should just resolve any conflicts by picking the changes that reflect your changeset and work as expected still with your updates.
> 
> <img alt="Screenshot 2023-05-22 at 10 30 03 AM" width="424" src="https://user-images.githubusercontent.com/8602661/239972118-d52f8f1e-0e91-4dc8-8b5a-c5eed3967c6c.png">

Looks like I don't have the proper access for this
![Screen Shot 2023-05-22 at 19 11 12](https://github.com/ArtBlocks/artblocks-engine-react/assets/14253814/c37b6003-2bbc-44a7-8eea-2ef9c81d8b93)

In other news - I have updated the readme to explain the project setup and configuration (also added a vercel deployment button), along with some other minor fixes. One major change is I added a new /user route to show the owned tokens for a given wallet. Hope that's okay!

jakerockland commented:
> > If you click on the "Use the command line" hyperlink that I believe you should see under the conflicts message, it should walk you through how to pull down the upstream changes into your branch, and then from there you should just resolve any conflicts by picking the changes that reflect your changeset and work as expected still with your updates.
> > <img alt="Screenshot 2023-05-22 at 10 30 03 AM" width="424" src="https://user-images.githubusercontent.com/8602661/239972118-d52f8f1e-0e91-4dc8-8b5a-c5eed3967c6c.png">
> 
> Looks like I don't have the proper access for this ![Screen Shot 2023-05-22 at 19 11 12](https://user-images.githubusercontent.com/14253814/240065276-c37b6003-2bbc-44a7-8eea-2ef9c81d8b93.png)

Hrm, any chance these instructions work? If not, I'm happy to try to resolve these on our end pre-merge.

<img width="833" alt="Screenshot 2023-05-22 at 5 39 37 PM" src="https://github.com/ArtBlocks/artblocks-engine-react/assets/8602661/484c0d28-520f-460e-b57b-771f0fb0fca1">


> In other news - I have updated the readme to explain the project setup and configuration (also added a vercel deployment button), along with some other minor fixes. One major change is I added a new /user route to show the owned tokens for a given wallet. Hope that's okay!

That is better than OK, its fantastic!

---

@matthewjacobson additional minters deployed for you [here](https://github.com/ArtBlocks/artblocks-engine-react/pull/10):

- https://github.com/ArtBlocks/artblocks-contracts/blob/2a63f35b9ae582ce07ab8ed89942fa4c37e7cc57/packages/contracts/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md#minter-deployment-2
- https://github.com/ArtBlocks/artblocks-contracts/blob/2a63f35b9ae582ce07ab8ed89942fa4c37e7cc57/packages/contracts/deployments/engine/V3/partners/plottables/DEPLOYMENTS.md#minter-deployment-3

matthewjacobson commented:
@jakerockland looks like my upstream rebase was successful :)

Thanks for those new contracts - I'll get to work on integrating them this week!

matthewjacobson commented:
@jakerockland integration was quicker than I thought it would be :) I have added support for the MinterDAExpV4 and MinterDAExpSettlementV1 minters. I ran test auctions for each but happy to set up more if you or the team would like to test it out for yourselves. Otherwise I think I am good to merge on my side!

jakerockland commented:
Amazing 🙏 🙏  🚢 🚢
